### ISSUE TITLE
Additional documentation/examples

### ISSUE THREAD

**Issue by nbdelorme** [ChrisLahaye/pm2-messages#4]
This appears to me like the best solution I've encountered so far for gathering prom-metrics from a clustered app in pm2. Unfortunately I can't get it to work. Could additional documentation/examples please be provided?

I run the following code against 4 clustered instances of an app hosted in pm2 4.4.0
```
import pm2mes from "pm2-messages";
import * as prom from "prom-client";

prom.collectDefaultMetrics();
pm2mes.onMessage("get_prom_register", () => prom.register.getMetricsAsJSON());

const foo = async (callback) => {
  const metricsArr = await pm2mes.getMessages("get_prom_register");
  console.log(metricsArr);
  callback(prom.AggregatorRegistry.aggregate(metricsArr).metrics());
};

foo((o) => console.log(o));
```

the result is ....
```
> pm2-prom-client@0.0.1 start /work/csgp/pm2-prom-client
> node ./dist/index.js

[]
```

Any hints as to why metricsArr is empty?

Regards

**Comment by ChrisLahaye** [comment]
@nbdelorme could you share the output of `pm2 list`?

**Comment by nbdelorme** [comment]
Thanks for getting back so quickly.

**pm2-messages version**
  "dependencies": {
    "@types/node": "^14.0.1",
    "pm2-messages": "^2.0.0",
    "prom-client": "^12.0.0",
    "typescript": "^3.9.2"
  }

**pm2  list**
![image](https://user-images.githubusercontent.com/5318775/82216239-aec0ea80-98e6-11ea-9146-8ba019ed453f.png)

**curl /metrics**
curl -z https://localhost/metrics
# HELP process_cpu_user_seconds_total Total user CPU time spent in seconds.
# TYPE process_cpu_user_seconds_total counter
process_cpu_user_seconds_total 1087.7053280000034 1589807215305
....

**Comment by ChrisLahaye** [comment]
@nbdelorme You are getting an empty array because none of the processes (including the active process) are managed by pm2. So using this library via `node` does not work. 

Currently the selection of processes to receive data from are determined by filtering (with as callback function the configuration option `filter`) the list of processes managed by PM2. What I can do is add another configuration option `includeSelf` with as value an enum:
- IfPm2 (default), to include the active process if its managed by PM2 and selected by the filter;
- IfNoPm2, to include the active process if it is not managed by PM2, i.e. when using `node`;
- Always, to always include the active process;
- Never, to never include the active process.

Any suggestions?

**Comment by nbdelorme** [comment]
@ChrisLahaye what do you mean by an app managed vs. not managed by pm2? Also, what do you mean by "the active process" (as in one single process)?

Apologies for my lack of understanding. I've been using pm2 in prod for years yet am not familiar with these concepts.

I am more than thankful you are suggesting a solution. I do not have enough data to make an informed suggestion.

**Comment by ChrisLahaye** [comment]
@nbdelorme The code you initially provided currently won't work because the process that is executing the code (i.e. the active process) is not running in PM2 (i.e. is not managed by PM2), this because you are using `node` instead of `pm2 start`. The selection of processes of which data needs to be received is determined by retrieving and filtering the list of processes running in PM2, which in your case is empy. The solution I suggested will allow some mutations on this list even though you are not using PM2. I will start implementing the solution and once done I will let you know which configuration to use.

In

**Comment by ChrisLahaye** [comment]
@nbdelorme Can you try the latest code on GitHub with the configuration option `includeSelfIfManaged`.

1. git clone https://github.com/ChrisLahaye/pm2-messages
2. yarn build
3. yarn link
4. cd YourProject; yarn link pm2-messages

Then use as follows:
```
await pm2mes.getMessages("get_prom_register", undefined, { includeSelfIfUnmanaged: true });
```

Now it should request the message from the active process when running outside PM2. Please confirm if everything works and I will release on npmjs.

**Comment by nbdelorme** [comment]
@ChrisLahaye apologies for not understanding sooner that the metrics collecting app should also be managed by pm2. It seems obvious now.

![image](https://user-images.githubusercontent.com/5318775/82448659-08a4ea00-9a78-11ea-8ee1-abee2f11ba6b.png)

Your new un-managed-by-pm2 solution might work just as well, but I will continue to attempt the managed solution.

**Client App**
Application _app_ now has the following code and I am able to view metrics on the /metrics route.
```
const pm2mes = require("pm2-messages");
const prom = require("prom-client");
const collectDefaultMetrics = prom.collectDefaultMetrics;
pm2mes.onMessage("get_prom_register", () => prom.register.getMetricsAsJSON());
```

**Metrics Collector**
Application _metrics_ should collect metrics from all instances of application _app_.
```
import pm2mes from "pm2-messages";
import * as prom from "prom-client";
let http = require("http");

//prom.collectDefaultMetrics(); /* 1 */
//pm2mes.onMessage("get_prom_register", () => prom.register.getMetricsAsJSON()); /* 2 */

http
  .createServer(async (req, res) => {
    //const metricsArr = await pm2mes.getMessages("get_prom_register"); /* 3 */
    const metricsArr = await pm2mes.getMessages(
      "get_prom_register",
      undefined,
      { filter: (process) => process.name === "app", timeout: 100 }
    );
    const data = prom.AggregatorRegistry.aggregate(metricsArr).metrics();
    return res.end(data);
  })
  .listen(8080);
```

**Ouput/Error**
```
You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:
AssertionError [ERR_ASSERTION]: Handler for get_prom_register not attached or not a function
    at Object.getMessages (/work/pm2-prom-client/node_modules/pm2-messages/src/index.ts:76:10)
    at Server.<anonymous> (/work/pm2-prom-client/src/index.ts:24:37)
    at Server.emit (events.js:310:20)
    at Server.<anonymous> (/opt/node/v12.16.3/lib/node_modules/pm2/node_modules/@pm2/io/src/metrics/httpMetrics.ts:181:25)
    at parserOnIncoming (_http_server.js:786:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:119:17)
```

If lines 1, 2, and 3 are used then prom-client data is returned for the _metrics_ application itself. I conclude that things are generally working. The only missing piece now is that a pm2mes.getMessages filter for process.name "app" results in a (unhandled) Promise rejection.

I will continue to experiment and debug. Any hints?

Regards.

**Comment by ChrisLahaye** [comment]
@nbdelorme This is because your metrics app does not have a handler for the topic `get_prom_register`. This is currently required. However, I see from your use case that making such an assumption is wrong. Therefore, I will remove this check. 

Not many people are using this library but I am using this in production myself and I am determined to maintain this project, so I appreciate your feedback!

**Comment by ChrisLahaye** [comment]
@nbdelorme I have removed the check. Could you please test the code in `master` and let me know whether you still have any issues. Looking forward to hear from you.

**Comment by nbdelorme** [comment]
Eureka! Along with your latest change, the following client app (express server) code work with the metrics collector code shared previously. 

@ChrisLahaye pm2-messages is fantastic.

```
const pm2mes = require("pm2-messages");
const prom = require("prom-client");
const collectDefaultMetrics = prom.collectDefaultMetrics;
const Registry = prom.Registry;
const register = new Registry();

collectDefaultMetrics({
    register: register,
    timeout: 5000,
    prefix: "app_",
});
pm2mes.onMessage("get_prom_register", () => register.getMetricsAsJSON());

app.use("/metrics", (request, response) => {
    response.set("Content-Type", register.contentType);
    response.send(register.metrics());
});
```

One bug to report. If multiple pm2 applications with the same name exist (e.g. app) and at least one is in stopped state then pm2mes.getMessages fails with the following.

![image](https://user-images.githubusercontent.com/5318775/82513761-e5168980-9ae1-11ea-8087-7a0cfda1b962.png)

```
You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:
Error: Process with ID <7> offline.
    at Object.God.logAndGenerateError (/opt/node/v12.16.3/lib/node_modules/pm2/lib/God/Methods.js:39:12)
    at God.sendDataToProcessId (/opt/node/v12.16.3/lib/node_modules/pm2/lib/God/ActionMethods.js:717:21)
    at Server.onmessage (/opt/node/v12.16.3/lib/node_modules/pm2/node_modules/pm2-axon-rpc/lib/server.js:104:6)
    at RepSocket.emit (events.js:310:20)
    at RepSocket.EventEmitter.emit (domain.js:482:12)
    at Parser.<anonymous> (/opt/node/v12.16.3/lib/node_modules/pm2/node_modules/pm2-axon/lib/sockets/rep.js:51:15)
    at Parser.emit (events.js:310:20)
    at Parser.EventEmitter.emit (domain.js:505:15)
    at Parser._write (/opt/node/v12.16.3/lib/node_modules/pm2/node_modules/amp/lib/stream.js:91:16)
    at doWrite (_stream_writable.js:403:12)
```

**Comment by ChrisLahaye** [comment]
The console message is shown by node since you don't capture any exception thrown by the getMessages function.

This exception is correct and the intended behavior. In your case you may want to extend the filter to only include online processes, so you can try adding process.pm2_env.status === 'online' according to the documentation on process descriptions at https://pm2.keymetrics.io/docs/usage/pm2-api/. Just check the available properties for the first parameter of the filter.

**Comment by nbdelorme** [comment]
process.pm2_env.status === 'online' worked as expected. Looking forward to release 3.x in npm registry.

Thanks @ChrisLahaye  for your support. I'm socializing this with my devops team but so far things look quite promising. I'll see if I can get you a few stars...

You may consider this "issue" resolved. This thread will certainly serve as documentation.

**Comment by ChrisLahaye** [comment]
@nbdelorme I just released 3.0.1 on npmjs. Feel free to reach out again if you encounter any issues.
